File: dynamic_programming/recursive_longest_common_subsequence.py
def common_lcs(x, y, idx_of_start, idx_of_end):
    if idx_of_start == 0 and idx_of_end == 0:
        result = 1 if x[idx_of_start] == y[idx_of_end] else 0
    elif idx_of_start == 0:
        result = 1 if x[idx_of_start] == y[idx_of_end] \
            else common_lcs(x, y, idx_of_start, idx_of_end - 1)
    elif idx_of_end == 0:
        result = 1 if x[idx_of_start] == y[idx_of_end] \
            else common_lcs(x, y, idx_of_start - 1, idx_of_end)
    else:
        if x[idx_of_start] == y[idx_of_end]:
            result = common_lcs(x, y, idx_of_start - 1, idx_of_end) + 1
        else:
            result = max(common_lcs(x, y, idx_of_start - 1, idx_of_end), \
                common_lcs(x, y, idx_of_start, idx_of_end - 1))

    return result


if __name__ == '__main__':
    x = 'ABCBDAB'
    y = 'BDCABA'

    for i in range(len(x)):
        for j in range(len(y)):
            result = common_lcs(x, y, i, j)

            print(f"i = {i}, j = {j}, X = {x[:i + 1]}, Y = {y[:j + 1]}, result = {result}")

File: dynamic_programming/recursive_longest_common_subsequence.go
//go:build ignore

package main

import "fmt"

func common_lcs(x, y string, i, j int) int {
	if i == 0 && j == 0 {
		if x[i] == y[j] {
			return 1
		} else {
			return 0
		}
	} else if i == 0 {
		if x[i] == y[j] {
			return 1
		} else {
			return common_lcs(x, y, i, j-1)
		}
	} else if j == 0 {
		if x[i] == y[j] {
			return 1
		} else {
			return common_lcs(x, y, i-1, j)
		}
	} else {
		if x[i] == y[j] {
			return common_lcs(x, y, i-1, j-1) + 1
		} else {
			if common_lcs(x, y, i, j-1) < common_lcs(x, y, i-1, j) {
				return common_lcs(x, y, i-1, j)
			} else {
				return common_lcs(x, y, i, j-1)
			}
		}
	}
}

func main() {
	x := "ABCBDAB"
	y := "BDCABA"

	for i := 0; i < len(x); i += 1 {
		fmt.Printf("--- i = %d ---\n", i)
		for j := 0; j < len(y); j += 1 {
			result := common_lcs(x, y, i, j)

			fmt.Printf("(i, j) = (%d, %d) / result = %d\n", i, j, result)
		}
	}
}
